.github/workflows/rebuilder.py

```diff
@@ -0,0 +1,300 @@
+import argparse
+import datetime
+import json
+import logging
+import re
+from typing import Set
+
+import copr.v3
+import dnf
+import hawkey
+
+
+def load_tests(loader, tests, ignore):
+    """We want unittest to pick up all of our doctests
+
+    See https://docs.python.org/3/library/unittest.html#load-tests-protocol
+    See https://stackoverflow.com/a/27171468
+    """
+    import doctest
+
+    tests.addTests(doctest.DocTestSuite())
+    return tests
+
+
+def filter_llvm_pkgs(pkgs: set[str]) -> set[str]:
+    """Filters out LLVM packages and returns the rest.
+
+    Args:
+        pkgs (set[str]): List of package names
+
+    Returns:
+        set[str]: List of package names without LLVM packages
+
+    Example:
+
+    >>> pkgs={'firefox', 'llvm99' 'libreoffice', 'clang18', 'compiler-rt'}
+    >>> filtered=list(filter_llvm_pkgs(pkgs))
+    >>> filtered.sort()
+    >>> print(filtered)
+    ['firefox', 'libreoffice']
+
+    """
+    llvm_pkgs = {
+        "llvm",
+        "clang",
+        "llvm-bolt",
+        "libomp",
+        "compiler-rt",
+        "lld",
+        "lldb",
+        "polly",
+        "libcxx",
+        "libclc",
+        "flang",
+        "mlir",
+    }
+    llvm_pkg_pattern = rf"({'|'.join(llvm_pkgs)})[0-9]*$"
+    return {pkg for pkg in pkgs if not re.match(llvm_pkg_pattern, pkg)}
+
+
+def get_exclusions() -> set[str]:
+    """
+    This returns a list of packages we don't want to test.
+    """
+    return set()
+
+
+def get_pkgs(exclusions: set[str]) -> set[set]:
+    base = dnf.Base()
+    conf = base.conf
+    for c in "AppStream", "BaseOS", "CRB", "Extras":
+        base.repos.add_new_repo(
+            f"{c}-source",
+            conf,
+            baseurl=[
+                f"https://odcs.fedoraproject.org/composes/production/latest-Fedora-ELN/compose/{c}/source/tree/"
+            ],
+        )
+    repos = base.repos.get_matching("*")
+    repos.disable()
+    repos = base.repos.get_matching("*-source*")
+    repos.enable()
+
+    base.fill_sack()
+    q = base.sack.query(flags=hawkey.IGNORE_MODULAR_EXCLUDES)
+    q = q.available()
+    q = q.filter(requires=["clang", "gcc", "gcc-c++"])
+    pkgs = [p.name for p in list(q)]
+    return filter_llvm_pkgs(set(pkgs)) - exclusions
+
+
+def get_builds_from_copr(
+    project_owner: str, project_name: str, copr_client: copr.v3.Client
+) -> list[dict]:
+    return copr_client.package_proxy.get_list(
+        project_owner,
+        project_name,
+        with_latest_succeeded_build=True,
+        with_latest_build=True,
+    )
+
+
+def get_monthly_rebuild_packages(pkgs: set[str], copr_pkgs: list[dict]) -> set[str]:
+    for p in copr_pkgs:
+        latest_succeeded = p["builds"]["latest_succeeded"]
+        latest = p["builds"]["latest"]
+        if p["name"] not in pkgs:
+            continue
+        if not latest_succeeded:
+            pkgs.discard(p["name"])
```

Comment:
I wonder if we should have a small list of important packages that should never get removed even if they failed last time, e.g. qemu.

Reply:
I was planning to add tmt tests for important packages, like qemu, so that they get tested every day, not just periodically with the mass rebuild, but I agree, I think it does make sense to ensure they always get tested by this script too.